rfcs/003-double-entry-ledger.md

```diff
@@ -0,0 +1,122 @@
+Collective. To address this difference, the Payment Method needs to
+have a ledger as well.
+
+### Create separate transactions for fees
+
+#### Use case example
+
+Given a transaction that represents moving $5 dollars from User to
+Collective with 5% of platform fee, 10% of host fee and 2.9%+30c of
+payment processor fee.
+
+##### How it is currently stored
+|            | User Ledger | Collective Ledger |
+|------------|------------:|------------------:|
+| Type       |       DEBIT |            CREDIT |
+| Amount     |       -4075 |              5000 |
+| Host Fee   |        -500 |              -500 |
+| Plat Fee   |        -250 |              -250 |
+| PP Fee     |        -175 |              -175 |
+| Net Amount |       -5000 |              4075 |
+
+##### How it should look like after this change
+
+|        |  User | Collective | Collective | Platform | Collective | Host | Collective | Payment Method |
+|--------|------:|-----------:|-----------:|---------:|-----------:|-----:|-----------:|---------------:|
+| Amount | -5000 |       5000 |       -250 |      250 |       -500 |  500 |       -175 |            175 |
+
+##### Where did the net amount go
+
+The net amount won't be a static value in the database anymore. The
+next section introduces an API that provides core operations to the
+ledger including calculate (& cache) the balance, getting fees etc.
+
+## Ledger API (`libocledger`)
+
+The Ledger API will now provide the following tools:
+
+ * `Number balance(Number CollectiveId)`
+
+   Return the amount of funds in a ledger after summing up all the
+   transactions.
+
+   ```javascript
+   > libledger.balance(665)
+   4325
+   ```
+
+ * `Object<String,Number> summary(String TransactionGroup)`
+
+   Return an object describing all the rows related to a single
+   transaction.
+
+   ```javascript
+   > libledger.summary('d3e36baa-69a2-4f6e-ac44-f93981e51c97')
+   { amount: 500, sumfees: 95, netAmount: 405 }
+   ```
+
+ * `Object<String,Number> fees(String TransactionGroup)`
+ 
+   Return an `Object` with keys containing the names of the fees and
+   their respective values.
+
+   ```javascript
+   > libledger.fees('d3e36baa-69a2-4f6e-ac44-f93981e51c97')
+   { platformFeeInHostCurrency: -25, hostFeeInHostCurrency: -25, paymentProcessorFeeInHostCurrency: -45 }
+   ```
+
+ * `Object<String,Number> rows(String TransactionGroup)`
+
+   Return an `Object` describing all the rows related to a single
+   transaction.
+
+   ```javascript
+   > libledger.rows('d3e36baa-69a2-4f6e-ac44-f93981e51c97')
+   [
+     { fromCollective: 10165, collective: 58, amount: -5000 }, // from User to Collective
+     { fromCollective: 58, collective: 1, amount: -250 },      // from Collective to Platform
+     { fromCollective: 58, collective: 11004, amount: -250 },  // from Collective to Host
+     { fromCollective: 58, collective: 90000, amount: -175 },  // from Collective to Payment Processor
+   ]
```

Comment:
👍 for creating a library to handle this. Would be nice to see the library handle multiple currencies as well.

Reply:
Yeah, I'll give some more thought to how the currencies should be handled in general and then think of which methods could wrap up these currency operations!